lib/logflare/backends.ex

```diff
@@ -169,25 +181,86 @@ defmodule Logflare.Backends do
+
+  TODO: Perform syncronous parsing and validation of log event params.
+
+  Once this function returns `:ok`, the events get dispatched to respective backend adaptor portions of the pipeline to be further processed.
   """
   @type log_param :: map()
   @spec ingest_logs([log_param()], Source.t()) :: :ok
-  def ingest_logs(log_events, source) do
-    via = via_source(source, :buffer)
-    Buffer.add_many(MemoryBuffer, via, log_events)
+  def ingest_logs(event_params, source) do
+    :ok = Source.Supervisor.ensure_started(source)
+    {log_events, errors} = split_valid_events(source, event_params)
+    count = Enum.count(log_events)
+    increment_counters(source, count)
+    maybe_broadcast_and_route(source, log_events)
+    RecentLogsServer.push(source, log_events)
+    dispatch_to_backends(source, log_events)
+    if Enum.empty?(errors), do: {:ok, count}, else: {:error, errors}
+  end
+
+  defp split_valid_events(source, event_params) do
+    event_params
+    |> Enum.reduce({[], []}, fn param, {events, errors} ->
+      le =
+        param
+        |> case do
+          %LogEvent{} = le ->
+            le
+
+          param ->
+            LogEvent.make(param, %{source: source})
+        end
+        |> Logs.maybe_mark_le_dropped_by_lql()
+        |> LogEvent.apply_custom_event_message()
+
+      cond do
+        le.drop ->
+          do_telemetry(:drop, le)
+          {events, errors}
+
+        le.valid == false ->
+          do_telemetry(:invalid, le)
+          {events, errors}
+
+        le.pipeline_error ->
+          {events, [le.pipeline_error.message | errors]}
+
+        le.valid ->
+          {[le | events], errors}
+      end
+    end)
+  end
+
+  defp increment_counters(source, count) do
+    Sources.Counters.increment(source.token, count)
+    SystemMetrics.AllLogsLogged.increment(:total_logs_logged, count)
     :ok
   end
 
-  @doc """
-  Dispatch log events to a given source backend.
-  It requires the source supervisor and registry to be running.
-  For internal use only, should not be called outside of the `Logflare` namespace.
-  """
-  def dispatch_ingest(log_events, source) do
+  defp maybe_broadcast_and_route(source, log_events) do
+    Logflare.Utils.Tasks.start_child(fn ->
+      if source.metrics.avg < 5 do
+        Source.ChannelTopics.broadcast_new(log_events)
+      end
+
+      SourceRouting.route_to_sinks_and_ingest(log_events)
+    end)
+
+    :ok
+  end
+
+  defp dispatch_to_backends(source, log_events) do
     Registry.dispatch(SourceDispatcher, source.id, fn entries ->
-      for {pid, {adaptor_module, :ingest}} <- entries do
+      for {pid, mfa} <- entries do
         # TODO: spawn tasks to do this concurrently
-        adaptor_module.ingest(pid, log_events)
+        case mfa do
+          {adaptor_module, :ingest, [_ | _] = opts} ->
+            adaptor_module.ingest(pid, log_events, opts)
+
+          {adaptor_module, :ingest} ->
+            adaptor_module.ingest(pid, log_events)
+        end
```

Comment:
I think that this could be handled in adaptor behaviour. In general I am not much of a fan of doing dynamic function calls outside of the behaviour that defines them as callbacks.

Reply:
ok refactored this away, added a Backends.

```
    :ok = Backends.register_backend_for_ingest_dispatch(__MODULE__, {source, backend})
```
this function will now add in the options automatically and handle the registry registration